### PR TITLE
ARM Translator

### PR DESCRIPTION
**Pull request by asimonetta**
ARM Translator

### CODE REVIEW

**Comment by codemanyak** [comment]
Hello Alessandro,
might it be that you forgot to check in the ARMGenerator.java file and some other stuff? I can only see two (uncommented) new flags in Element.java in the pull request...

**Comment by asimonetta** [comment]
now all is ok, NetBeans has committed all updates!

File: src/lu/fisch/utils/StringList.java [hunk 1]
```diff
@@ -1352,6 +1352,25 @@ public int replaceAllBetween(String _stringOld, String _stringNew, boolean _matc
     }
     // END KGU#129 2016-01-08
 
+
+	// ARM
+	public int replaceIfContains(String _stringOld, String _stringNew){
```
Comment (codemanyak) [inline src/lu/fisch/utils/StringList.java:1357]: What is the result meant to be? (A complete Javadoc comment would be nice anyway.) Sensible meanings would be a) the total number of string replacements (difficult to get) or b) the number of actually modified member strings. Unfortunately the actual result is none of them but simply the number of elements the loop has run through - more precisely = this.count() while the prepared index range limitation is only a hidden internal fragment.

File: src/lu/fisch/structorizer/gui/Diagram.java [hunk 2]
```diff
@@ -9119,7 +9120,20 @@ protected void setCodePreviewTooltip() {
 		}
 	}
 	// END KGU#705 2019-09-24
-
+/*
```
Comment (codemanyak) [inline src/lu/fisch/structorizer/gui/Diagram.java:9123]: This is to be changed into a Javadoc comment (two asterisks)

File: src/lu/fisch/structorizer/gui/Diagram.java [hunk 3]
```diff
@@ -9119,7 +9120,20 @@ protected void setCodePreviewTooltip() {
...
+* Show arm visual
+* @param _arm - true to switch in arm mode
+*/
+public void setOperationArmVisual(boolean _arm) {
+Element.ARM_VISUAL = _arm;
```
Comment (codemanyak) [inline src/lu/fisch/structorizer/gui/Diagram.java:9128]: Indentation is to be fixed...

File: src/lu/fisch/structorizer/generators/ArmGenerator.java [hunk 4]
```diff
@@ -0,0 +1,1363 @@
...
+
+        addCode(c, "\t\t", isDisabled);
+    }
+
+    /**
+     * Transforms a string in a char array
+     **/
+
+    public void generateString(String line, boolean isDisabled) {
+        line = line.replace(" ", "");
```
Comment (codemanyak) [inline src/lu/fisch/structorizer/generators/ArmGenerator.java:951]: Wouldn't this cause harm to string literals containing blanks?

File: src/lu/fisch/structorizer/generators/ArmGenerator.java [hunk 5]
```diff
@@ -0,0 +1,1363 @@
...
+            mode = "AInitialization";
+        else if (address.matcher(line).matches())
+            mode = "Address";
+        else if (isArmInstruction(line))
+            mode = "Instruction";
+        return mode;
+    }
+
+    // Changes variables to a register
+
```
Comment (codemanyak) [inline src/lu/fisch/structorizer/generators/ArmGenerator.java]: As it seems, this method variant isn't used anywhere.

File: src/lu/fisch/structorizer/generators/ArmGenerator.java [hunk 6]
```diff
@@ -0,0 +1,1363 @@
...
+            }
+            if (lines[i].startsWith("end_") && lines[i + 1].startsWith("end_")) {
+                code.replaceAll(lines[i], lines[i] + ":");
+                code.replaceIfContains(lines[i] + ":", "");
+                code.replaceIfContains(lines[i].replace(difference[0], ""), lines[i + 1].replace(difference[0], "").replace("\"]", ""));
+            }
+        }
+    }
+
+    public void appendTop(String line) {
```
Comment (codemanyak) [inline src/lu/fisch/structorizer/generators/ArmGenerator.java:1200]: This inserts *after* the first line of code!
**And** it will confuse the code preview!

File: src/lu/fisch/structorizer/generators/ArmGenerator.java [hunk 7]
```diff
@@ -0,0 +1,1363 @@
...
+
+    /*
+    Key contains the variable, oldKey contains already assigned variables
+    */
+    public void setVariables(String key, String oldKey) {
+        String Rl;
+        if (!key.equals("")) {
+            Rl = notUsedRegister(key + oldKey);
+            if (Rl == null) {
+                appendComment("No Free Registers for variable: " + key, "\t\t");
```
Comment (codemanyak) [inline src/lu/fisch/structorizer/generators/ArmGenerator.java:1322]: May `"\t\t"` be replaced by `getIndent()`?

File: src/lu/fisch/structorizer/generators/ArmGenerator.java [hunk 8]
```diff
@@ -0,0 +1,1363 @@
...
+                return;
+            }
+
+            if (!variables.containsKey(key))
```
Comment (codemanyak) [inline src/lu/fisch/structorizer/generators/ArmGenerator.java]: See remark for method `setVariables(String)`.

File: src/lu/fisch/structorizer/generators/ArmGenerator.java [hunk 9]
```diff
@@ -0,0 +1,1363 @@
...
+                    setVariables(key, oldKey);
+                key = variables.get(key);
+            }
+        }
+
+        return key;
+    }
+
+    // A list of prefixes that we can use for the debugger
+
```
Comment (codemanyak) [inline src/lu/fisch/structorizer/generators/ArmGenerator.java]: It is is to be used for the debugger then the method should rather be a static method, perhaps.

File: src/lu/fisch/utils/StringList.java [hunk 10]
```diff
@@ -16,50 +16,48 @@
 
     You should have received a copy of the GNU General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.
-*/
-
+ */
 package lu.fisch.utils;
```
Comment (codemanyak) [inline src/lu/fisch/utils/StringList.java]: Oops! The complete reformatting of this file was not so good an idea, I am afraid: Makes it hard to track the substantial differences on GitHub... (But there are other tools, o course.)

Comment (codemanyak) [reply src/lu/fisch/utils/StringList.java]: Okay. The Change history and some modification comments will have to be restored. In the event, the file gained a consistent code style. Method comment of `replaceIfContained` still refers to a no longer existent result value.

File: src/lu/fisch/structorizer/gui/Menu.java [hunk 11]
```diff
@@ -1365,7 +1366,9 @@ public void actionPerformed(ActionEvent event)
 		menuDiagram.add(menuDiagramExpand);
 		menuDiagramExpand.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_ADD, 0));
 		menuDiagramExpand.addActionListener(new ActionListener() { public void actionPerformed(ActionEvent event) { diagram.expandNSD(); doButtons(); } } );
-
+//ARM
```
Comment (codemanyak) [inline src/lu/fisch/structorizer/gui/Menu.java]: I think for the next version(s) - while there are no modifications to the debugger and other stuff - this menu item will be replaced by a plugin-specific export option. Might be reactivated later.

File: src/lu/fisch/structorizer/gui/IconLoader.java [hunk 12]
```diff
@@ -238,7 +238,8 @@
 			"125_plug.png",			// introduced with 3.30-14 for #910
 			"126_info.png",			// is to replace the 017_Eye.png in the main toolbox
 			"127_merge.png",		// introduced with 3.30-16 for #915
-			"128_split.png"			// introduced with 3.30-16 for #915
+			"128_split.png",		// introduced with 3.30-16 for #915
+                        "129_arm.png"
```
Comment (codemanyak) [inline src/lu/fisch/structorizer/gui/IconLoader.java]: File name is likely to be renamed, most likely to "129_gnu.png". It may still be used for some ARM-related stuff but it shows just the gnu logo.

File: nbproject/private/private.xml [hunk 13]
```diff
@@ -4,8 +4,8 @@
     <editor-bookmarks xmlns="http://www.netbeans.org/ns/editor-bookmarks/2" lastBookmarkId="0"/>
     <open-files xmlns="http://www.netbeans.org/ns/projectui-open-files/2">
         <group>
-            <file>file:/Users/robertfisch/Programs/Structorizer/buildapp.xml</file>
-            <file>file:/Users/robertfisch/Programs/Structorizer/build.xml</file>
+            <file>file:/Users/ale/NetBeansProjects/Structorizer.Desktop/src/lu/fisch/structorizer/elements/Element.java</file>
+            <file>file:/Users/ale/NetBeansProjects/Structorizer.Desktop/src/lu/fisch/structorizer/gui/Diagram.java</file>
```
Comment (codemanyak) [inline nbproject/private/private.xml:8]: To adopt these changes might be harmful to the release process for Bob...

File: .gitignore [hunk 14]
```diff
@@ -1 +1,2 @@
 /bin
+/dist/
```
Comment (codemanyak) [inline .gitignore:2]: Might be harmful for the release script (?)

File: nbproject/private/private.properties [hunk 15]
```diff
@@ -1,7 +1,10 @@
 compile.on.save=false
 do.depend=false
 do.jar=true
+do.jlink=false
+file.reference.bsh-2.0b4.jar-1=/Users/ale/NetBeansProjects/Structorizer.Desktop/lib/bsh-2.0b6.jar
```
Comment (codemanyak) [inline nbproject/private/private.properties:5]: These changes might be a problem for Bob's release scripts.

File: nbproject/private/private.properties [hunk 16]
```diff
@@ -1,7 +1,10 @@
...
 javac.debug=true
 javadoc.preview=true
 jaxbwiz.endorsed.dirs=/Applications/NetBeans/NetBeans 6.8.app/Contents/Resources/NetBeans/ide12/modules/ext/jaxb/api
-user.properties.file=/Users/robertfisch/Library/Application Support/NetBeans/8.0.1/build.properties
+jlink.strip=false
+user.properties.file=/Users/ale/Library/Application Support/NetBeans/12.2/build.properties
```
Comment (codemanyak) [inline nbproject/private/private.properties:10]: As above. Bob should have a llok at it and possibly revert the changes if necessary.

**Review by codemanyak — COMMENTED** [review]
The reformatting of Diagram.java is hard stuff (it certainly improves code style but definitely breaks change trackability beyond this point and requires restoration of the change history and many modification markers). I am inclined to ask to revert it.

The changes to the nbproject files don't touch me (as I am working with Eclipse) but might compromise the release production at Bob's site. On the other hand, they are not too complex to be mended by Bob.

**Review by codemanyak — APPROVED** [review]
Adopted for run time checks
